rfcs/20200520-tensor-float-32.md

```diff
@@ -0,0 +1,80 @@
+# TensorFloat-32 in TensorFlow
+
+| Status        | Proposed                                             |
+:-------------- |:---------------------------------------------------- |
+| **RFC #**     | [247](https://github.com/tensorflow/community/pull/247) |
+| **Author(s)** | Reed Wanderman-Milne (reedwm@google.com)             |
+| **Sponsor**   | Sanjoy Das (sanjoy@google.com)                 |
+| **Updated**   | 2020-05-20                                           |
+
+## Objective
+
+Allow [TensorFloat-32](https://blogs.nvidia.com/blog/2020/05/14/tensorfloat-32-precision-format) to be used in TensorFlow to improve performance.
+
+## Motivation
+
+[NVIDIA Ampere](https://www.nvidia.com/en-us/data-center/nvidia-ampere-gpu-architecture/), an upcoming generation of NVIDIA GPUs announced at GTC 2020, introduces a new numeric format called TensorFloat-32, or TF32 for short.
+TF32 has the range of float32/bfloat16 (i.e. 8 bits of exponent) and the precision of fp16 (i.e. 10 bits of mantissa).
+It is not an in-memory format, but tensor cores natively support it as a computation format.
+TF32 should not be thought of as an in-memory dtype but instead a computation mode that increases performance and decreases numeric precision for certain float32 operations.
+NVIDIA has not found any cases where TF32 reduces the convergence of deep learning models.
+
+Upcoming versions of cuDNN, cuBLAS, and other CUDA libraries will expose a mode of execution that has float32 inputs and outputs, but internally truncates float32 to TF32 and uses tensor cores.  This is expected to be sufficiently accurate to reach the same convergence as the “full” float32 mode of execution but significantly faster.  Each element still takes four bytes, so there is still a memory and performance penalty compared to using float16 or bfloat16.
+
+As TF32 is only usable by tensor cores, it can only be used for matrix multiplications and other ops implemented in terms of matrix multiplications, such as convolutions. It is not used for pointwise ops or reductions.
+
+TF32 will benefit users who run float32 models on Ampere GPUs, so we need an API to allow these users to enable TF32. 
+
+## Design Proposal
+
+In TensorFlow, TF32 can be enabled for supported ops on Ampere GPUs with the following call:
+
+```python
+tf.config.allow_tensor_float_32_execution(True)
```

Comment:
Why not using the Keras mixed precision policy API?

Reply:
This affects ops outside Keras, so it shouldn't be under `tf.keras`. In a sense, TF32 is a form of mixed precision, as some ops use TF32 and others use float32. We could put it under `tf.mixed_precision`, but I think `tf.config` is better since tf32 should be though of as a mode, not a dtype.

Reply:
Also the mixed precision API mostly changes the dtype of tensors, while tf32 doesn't affect tensor dtype (afaict) just the dtype of accumulators inside ops.